source-pendo/source_pendo/api.py

```diff
@@ -0,0 +1,123 @@
+import json
+from logging import Logger
+from typing import AsyncGenerator, Iterable
+
+from estuary_cdk.capture.common import LogCursor
+from estuary_cdk.http import HTTPSession
+
+from .models import Event, EventResponse, Resource, Metadata
+
+API = "https://app.pendo.io/api/v1"
+
+
+async def fetch_resources(
+        http: HTTPSession,
+        entity: str,
+        log: Logger,
+) -> Iterable[Resource]:
+    url = f"{API}/{entity}"
+
+    response = json.loads(await http.request(log, url))
+
+    resources: list[Resource] = []
+    for id, resource in enumerate(response):
+        r = Resource.model_validate_json(json.dumps(resource))
+        r.meta_ = Resource.Meta(op="c", row_id=id)
+        resources.append(r)
+
+    log.info(f"Read {len(resources)} documents from stream {entity}.")
+
+    return resources
+
+
+async def fetch_metadata(
+        http: HTTPSession,
+        entity: str,
+        log: Logger,
+) -> Iterable[Metadata]:
+    url = f"{API}/metadata/schema/{entity}"
+
+    metadata = Metadata.model_validate_json(
+        await http.request(log, url)
+    )
+
+    metadata.meta_ = Metadata.Meta(op="c")
+
+    log.info(f"Read 1 document from stream {entity}.")
+
+    return [
+        metadata
+    ]
+
+
+async def fetch_events(
+        http: HTTPSession,
+        entity: str,
+        timestamp_field: str,
+        log: Logger,
+        log_cursor: LogCursor,
+) -> AsyncGenerator[Event | LogCursor, None]:
+    assert isinstance(log_cursor, int)
+    url = f"{API}/aggregation"
+    last_ts = log_cursor
+
+    # Pendo's event aggregation endpoints require a JSON body with various fields.
+    # See https://engageapi.pendo.io/#f42c416f-bdd6-4082-b86c-1be3fe8401bf. We use
+    # their filtering and sorting capabilities to only return documents we haven't
+    # seen yet in ascending order.
+    # Note: The Pendo API limits aggregations to 5 minute runtime or 4 GB output.
+    # If/when we run up against those limits, a different strategy like requesting
+    # smaller time windows will be required.
+    body = {
+        "response": {
+            "mimeType": "application/json"
+        },
+        "request": {
+            "pipeline": [
+                {
+                    "source": {
+                        entity: None,
+                        "timeSeries": {
+                            "period": "hourRange",
+                            "first": f"{last_ts}",
+                            "last": "now()"
```

Comment:
Main conceptual question here: For these aggregate events, does the API return "partial" hour aggregates? For example, if we request the value for the current hour and the current hour isn't complete, are we going to get an incomplete aggregation and then never see it again?

Another thought is that we should probably bound the size of these responses. If the configured start time is far in the past, I'd imagine we could end hitting API limits or OOMing the connector on the response size. Could we use `count` to get 24 hours at a time or something like that?

Reply:
My understanding from the docs & testing is that aggregates returned by the API are as up to date as they can be, and a timestamp field is used to indicate when that aggregation was last updated. So for the current hour, we'll get an "incomplete" aggregation. But if that aggregation is updated within the current hour, it will be returned in a later API response since it's "last updated" timestamp (either `lastTime` or `guideTimestamp`) will be updated. The filter in the request body makes sure we get all aggregations with a "last updated" timestamp after the last aggregate we saw.

To bound the response sizes, there's no option to specify a certain number of days. We can use `limit` to limit how many lines we get though. I'll test that out & try to empirically find a reasonable number.

Reply:
While digging into this, I learned that not all of the event endpoints are actually aggregates. For `pollEvents` and `guideEvents`, a document is a single, distinct event. For `pageEvents`, `featureEvents`, and `trackEvents`, a document is an aggregate of multiple events within an hour. I ended up handling events & aggregated events separately due to their differences.

For the aggregated events, I added another condition to the `filter` operation in the API request. Aggregates are included in the response if its `hour` (the bottom of the hour, like 6:00 if the aggregate was updated at 6:15) is less than the bottom of the current hour. So if the current time is 6:30, we won't get aggregates for the ongoing 6:00 hour, but we will for 5:00, 4:00, etc.

I also noticed that Pendo's API isn't updated in real-time; it took ~30 minutes for the past hour's aggregates to show up in API calls. So this strategy might not be _completely_ necessary right now, but it is the right way to do it & covers us in case the API becomes closer to real-time.